semgrep/tests/e2e/snapshots/test_check/test_nested_pattern_either_rule/results.json

```diff
@@ -53,25 +35,7 @@
         "lines": "    nested_patterns_func('bar', 2)",
         "message": "test",
         "metadata": {},
-        "metavars": {
-          "$X": {
-            "abstract_content": "'bar'",
-            "end": {
-              "col": 31,
-              "line": 3,
-              "offset": 83
-            },
-            "start": {
-              "col": 26,
-              "line": 3,
-              "offset": 78
-            },
-            "unique_id": {
-              "md5sum": "<masked in tests>",
-              "type": "AST"
-            }
-          }
-        },
+        "metavars": {},
```

Comment:
who those metavars disappear from results.json?

Reply:
oops, I didn't notice these were gone. I still don't know what's happening. `semgrep-core` returns correct results, just in a different order.

Reply:
I think I understand. semgrep "deduplicates" the matches it receives from semgrep-core. It considers duplicates matches that occur at the same location, regardless of what the metavariables captured, and this is order-dependent.

The rule file here is
```
rules:
- id: test
  message: test
  severity: WARNING
  languages:
  - ts
  - js
  patterns:
  - pattern-either:
    - pattern: nested_patterns_func('foo', ...)
    - pattern: nested_patterns_func('bar', ...)
  - pattern-either:
    - pattern: nested_patterns_func($X, 1, ...)
    - pattern: nested_patterns_func($X, 2, ...)
```
It can be simplified to
```
rules:
- id: test
  message: test
  severity: WARNING
  languages:
  - ts
  - js
  patterns:
  - pattern-either:
    - pattern: f(...)
    - pattern: f($X)
```
and the following target:
```
f(x)
```
semgrep-core returns 2 matches, one without a metavariable and one with `$X = 'x'`. Because they match at the same location, the first match is kept by semgrep and the second one is discarded.

Here are the old results without sorting in semgrep-core:
```
$ semgrep-core -config repro2.yaml repro2.js -json | ydump
{
  "matches": [
    {
      "rule_id": "test",
      "location": {
        "path": "repro2.js",
        "start": { "line": 1, "col": 1, "offset": 0 },
        "end": { "line": 1, "col": 5, "offset": 4 },
        "lines": []
      },
      "extra": {
        "message": "test",
        "metavars": {
          "$X": {
            "start": { "line": 1, "col": 3, "offset": 2 },
            "end": { "line": 1, "col": 4, "offset": 3 },
            "abstract_content": "x",
            "unique_id": {
              "type": "AST",
              "md5sum": "52a71841547db9cbd91c34927fc21f01"
            }
          }
        }
      }
    },
    {
      "rule_id": "test",
      "location": {
        "path": "repro2.js",
        "start": { "line": 1, "col": 1, "offset": 0 },
        "end": { "line": 1, "col": 5, "offset": 4 },
        "lines": []
      },
      "extra": { "message": "test", "metavars": {} }
    }
  ],
  "errors": [],
  "skipped": [],
  "stats": { "okfiles": 1, "errorfiles": 0 }
}

$ semgrep -c repro2.yaml repro2.js --json | ydump
Running 1 rules...
ran 1 rules on 1 files: 1 findings
{
  "errors": [],
  "results": [
    {
      "check_id": "test",
      "end": { "col": 5, "line": 1, "offset": 4 },
      "extra": {
        "is_ignored": false,
        "lines": "f(x)",
        "message": "test",
        "metadata": {},
        "metavars": {
          "$X": {
            "abstract_content": "x",
            "end": { "col": 4, "line": 1, "offset": 3 },
            "start": { "col": 3, "line": 1, "offset": 2 },
            "unique_id": {
              "md5sum": "52a71841547db9cbd91c34927fc21f01",
              "type": "AST"
            }
          }
        },
        "severity": "WARNING"
      },
      "path": "repro2.js",
      "start": { "col": 1, "line": 1, "offset": 0 }
    }
  ]
}
```
and the results now with sorting in semgrep-core:
```
$ semgrep-core -config repro2.yaml repro2.js -json | ydump
{
  "matches": [
    {
      "rule_id": "test",
      "location": {
        "path": "repro2.js",
        "start": { "line": 1, "col": 1, "offset": 0 },
        "end": { "line": 1, "col": 5, "offset": 4 },
        "lines": []
      },
      "extra": { "message": "test", "metavars": {} }
    },
    {
      "rule_id": "test",
      "location": {
        "path": "repro2.js",
        "start": { "line": 1, "col": 1, "offset": 0 },
        "end": { "line": 1, "col": 5, "offset": 4 },
        "lines": []
      },
      "extra": {
        "message": "test",
        "metavars": {
          "$X": {
            "start": { "line": 1, "col": 3, "offset": 2 },
            "end": { "line": 1, "col": 4, "offset": 3 },
            "abstract_content": "x",
            "unique_id": {
              "type": "AST",
              "md5sum": "52a71841547db9cbd91c34927fc21f01"
            }
          }
        }
      }
    }
  ],
  "errors": [],
  "skipped": [],
  "stats": { "okfiles": 1, "errorfiles": 0 }
}

$ semgrep -c repro2.yaml repro2.js --json | ydump
Running 1 rules...
ran 1 rules on 1 files: 1 findings
{
  "errors": [],
  "results": [
    {
      "check_id": "test",
      "end": { "col": 5, "line": 1, "offset": 4 },
      "extra": {
        "is_ignored": false,
        "lines": "f(x)",
        "message": "test",
        "metadata": {},
        "metavars": {},
        "severity": "WARNING"
      },
      "path": "repro2.js",
      "start": { "col": 1, "line": 1, "offset": 0 }
    }
  ]
}
```

Reply:
I don't know if we should do anything about this. It looks like proper behavior to me.

Reply:
ok, sounds good to me. Thanks for investigating.

Reply:
Not something that concerns this PR, but maybe Semgrep's deduplication should be deterministic and not depend on the order in which semgrep-core outputs the findings. cc @brendongo